lib/internal/modules/esm/resolve.js

```diff
@@ -978,6 +978,15 @@ function defaultResolve(specifier, context = {}) {
     }
   }
 
+  if (specifier === 'node:interop') {
+    const path = fileURLToPath(parsedParentURL);
+    const dir = dirname(path);
+    return {
+      url: `data:text/javascript,export const __filename=${JSONStringify(path)};export const __dirname=${JSONStringify(dir)}`,
```

Comment:
can this be implemented as a SyntheticModule?

Reply:
What would be the benefit of doing so? (Not opposed, just curious.)

Reply:
Cheaper on VM side / doesn't need to tie it to a poison-able cache key.

Reply:
I assume we could also implement these as functions, like:

```js
import { dirname } from 'node:interop'
const __dirname = dirname()
```

And then maybe we wouldn’t need to create them ahead of time. We could generate the code to be evaluated, and then evaluate it, only when the function is called.

Yes it’s slightly more annoying to type `dirname()` instead of `__dirname` or `dirname`, but if it improves performance then it’s probably worth it.

Reply:
I think a cached getter would be a "6 of one", but with better DX :) i think no additional implementation effort (just a choice of A or B).

Reply:
It isn't just CPU/Mem when having unused values here, `data:` will cache the source text in the V8 Context's Module cache and engage a full parser vs just piping values where SyntheticModule ~= a Map and avoiding cache growing.